modules/rte_mis_student_tracking/src/Controller/StudentTrackingLogsDownload.php

```diff
@@ -0,0 +1,151 @@
+use Symfony\Component\HttpFoundation\BinaryFileResponse;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpFoundation\ResponseHeaderBag;
+use Symfony\Component\HttpKernel\Exception\AccessDeniedHttpException;
+
+/**
+ * Class to export logs related to student tracking module.
+ */
+class StudentTrackingLogsDownload extends ControllerBase {
+
+  /**
+   * The file system.
+   *
+   * @var \Drupal\Core\File\FileSystemInterface
+   */
+  protected $fileSystem;
+
+  /**
+   * The file repository service under test.
+   *
+   * @var \Drupal\file\FileRepository
+   */
+  protected $fileRepository;
+
+  /**
+   * The tempstore factory.
+   *
+   * @var \Drupal\Core\TempStore\PrivateTempStoreFactory
+   */
+  protected $tempStoreFactory;
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container) {
+    return new static(
+      $container->get('file_system'),
+      $container->get('file.repository'),
+      $container->get('tempstore.private')
+    );
+  }
+
+  /**
+   * Constructs StudentTrackingLogsDownload object.
+   *
+   * @param \Drupal\Core\File\FileSystemInterface $file_system
+   *   The file system.
+   * @param \Drupal\file\FileRepositoryInterface $fileRepository
+   *   The file repository.
+   * @param \Drupal\Core\TempStore\PrivateTempStoreFactory $temp_store_factory
+   *   The tempstore factory.
+   */
+  public function __construct(FileSystemInterface $file_system, FileRepositoryInterface $fileRepository, PrivateTempStoreFactory $temp_store_factory) {
+    $this->fileSystem = $file_system;
+    $this->fileRepository = $fileRepository;
+    $this->tempStoreFactory = $temp_store_factory;
+  }
+
+  /**
+   * Export the logs from students bulk import.
+   */
+  public function getStudentsImportLogs(int $fid = 0) {
+    $file = $this->entityTypeManager()->getStorage('file')->load($fid);
+    if ($file instanceof FileInterface) {
+      if ($file->getOwnerId() !== $this->currentUser()->id()) {
+        throw new AccessDeniedHttpException('Cannot access the log.');
+      }
+      $destinationUri = 'public://student-import-logs';
+      $this->fileSystem->prepareDirectory($destinationUri, FileSystemInterface:: CREATE_DIRECTORY | FileSystemInterface::MODIFY_PERMISSIONS);
+      $newFile = $this->fileRepository->copy($file, $destinationUri, FileSystemInterface::EXISTS_RENAME);
+      if ($newFile instanceof FileInterface) {
+        $newFileUri = $this->fileSystem->realpath($newFile->getFileUri());
+        $spreadsheet = IOFactory::load($newFileUri);
+        // Make changes to the spreadsheet.
+        $sheet = $spreadsheet->getActiveSheet();
+        // Get the store collection.
+        $store = $this->tempStoreFactory->get('rte_mis_student_tracking');
+        $studentsImportLogs = $store->get('students_import_logs');
+        if (!empty($studentsImportLogs)) {
+          $sheet->getStyle('A1:M1')->applyFromArray([
+            'font' => [
```

Comment:
Looks like this code is getting repeated. Can't we create a helper function for this?

Reply:
This is needed only once I pushed duplicate code by mistake.